File: SDK/CentralService/dotnet/src/CentralService.Service/CentralServiceServiceClient.cs
using System;
using System.Collections.Generic;
using System.Net;
using CentralService.Service.Errors;
using CentralService.Service.Internal;
using CentralService.Service.Models;
using CentralService.Shared.Internal;

namespace CentralService.Service
{
    /// <summary>
    /// 提供服务注册、心跳与注销相关的中心服务客户端能力。
    /// </summary>
    public sealed class CentralServiceServiceClient : IDisposable
    {
        private readonly ICentralServiceTransport _transport;

        /// <summary>
        /// 使用指定配置初始化中心服务客户端。
        /// </summary>
        /// <param name="options">SDK 连接配置。</param>
        public CentralServiceServiceClient(CentralServiceSdkOptions options)
            : this(options, null)
        {
        }

        internal CentralServiceServiceClient(CentralServiceSdkOptions options, Func<DateTimeOffset>? utcNowProvider)
        {
            if (options == null) throw new ArgumentNullException("options");
            var endpoints = options.CreateTransportEndpoints();
            if (options.HttpMessageHandler != null)
            {
                _transport = new CentralServiceHttpClientMultiEndpointTransport(
                    endpoints,
                    options.Timeout,
                    options.HttpMessageHandler,
                    utcNowProvider);
            }
            else
            {
                _transport = new CentralServiceMultiEndpointTransport(
                    endpoints,
                    options.Timeout,
                    options.IgnoreSslErrors,
                    utcNowProvider);
            }
        }

        /// <summary>
        /// 向中心服务注册当前服务实例。
        /// </summary>
        public ServiceRegistrationResponse Register(ServiceRegistrationRequest request)
        {
            if (request == null) throw new ArgumentNullException("request");

            var transport = Send("POST", "/api/Service/register", CentralServiceJson.Serialize(request));
            if (!IsSuccess(transport.StatusCode))
            {
                throw CreateParsedError("POST", transport, CentralServiceErrorParser.Parse("POST", transport.Url, transport.StatusCode, transport.Body));
            }

            var parsed = CentralServiceJson.Deserialize<ApiResponse<ServiceRegistrationResponse>>(transport.Body);
            if (parsed == null)
            {
                throw CreateParsedError(
                    "POST",
                    transport,
                    new CentralServiceError(
                        transport.StatusCode,
                        "POST",
                        transport.Url,
                        CentralServiceErrorKind.Unknown,
                        "无法解析响应",
                        null,
                        string.Empty,
                        transport.Body));
            }

            if (!parsed.Success)
            {
                throw CreateParsedError(
                    "POST",
                    transport,
                    new CentralServiceError(
                        transport.StatusCode,
                        "POST",
                        transport.Url,
                        CentralServiceErrorKind.ApiResponse,
                        parsed.ErrorMessage,
                        parsed.ErrorCode,
                        parsed.ErrorKey,
                        transport.Body));
            }

            return parsed.Data;
        }

        /// <summary>
        /// 向中心服务上报指定服务实例的心跳。
        /// </summary>
        public void Heartbeat(string serviceId)
        {
            if (string.IsNullOrWhiteSpace(serviceId)) throw new ArgumentNullException("serviceId");

            var transport = Send(
                "POST",
                "/api/Service/heartbeat",
                CentralServiceJson.Serialize(new ServiceHeartbeatRequest { Id = serviceId }));

            if (!IsSuccess(transport.StatusCode))
            {
                throw CreateParsedError("POST", transport, CentralServiceErrorParser.Parse("POST", transport.Url, transport.StatusCode, transport.Body));
            }

            var parsed = CentralServiceJson.Deserialize<ApiResponse<object>>(transport.Body);
            if (parsed != null && !parsed.Success)
            {
                throw CreateParsedError(
                    "POST",
                    transport,
                    new CentralServiceError(
                        transport.StatusCode,
                        "POST",
                        transport.Url,
                        CentralServiceErrorKind.ApiResponse,
                        parsed.ErrorMessage,
                        parsed.ErrorCode,
                        parsed.ErrorKey,
                        transport.Body));
            }
        }

        /// <summary>
        /// 从中心服务注销指定服务实例。
        /// </summary>
        public void Deregister(string serviceId)
        {
            if (string.IsNullOrWhiteSpace(serviceId)) throw new ArgumentNullException("serviceId");

            var path = "/api/Service/deregister/" + Uri.EscapeDataString(serviceId);
            var transport = Send("DELETE", path, null);
            if (!IsSuccess(transport.StatusCode))
            {
                throw CreateParsedError("DELETE", transport, CentralServiceErrorParser.Parse("DELETE", transport.Url, transport.StatusCode, transport.Body));
            }

            var parsed = CentralServiceJson.Deserialize<ApiResponse<object>>(transport.Body);
            if (parsed != null && !parsed.Success)
            {
                throw CreateParsedError(
                    "DELETE",
                    transport,
                    new CentralServiceError(
                        transport.StatusCode,
                        "DELETE",
                        transport.Url,
                        CentralServiceErrorKind.ApiResponse,
                        parsed.ErrorMessage,
                        parsed.ErrorCode,
                        parsed.ErrorKey,
                        transport.Body));
            }
        }

        /// <summary>
        /// 释放客户端持有的资源。
        /// </summary>
        public void Dispose()
        {
        }

        private CentralServiceTransportResult Send(string method, string path, string? jsonBody)
        {
            try
            {
                return _transport.Send(method, path, jsonBody);
            }
            catch (CentralServiceTransportExhaustedException ex)
            {
                throw new CentralServiceException(
                    new CentralServiceError(
                        HttpStatusCode.ServiceUnavailable,
                        method,
                        string.IsNullOrWhiteSpace(ex.LastUrl) ? path : ex.LastUrl,
                        CentralServiceErrorKind.Transport,
                        ex.Message,
                        null,
                        string.Empty,
                        ex.RawDetail));
            }
        }

        private static CentralServiceException CreateParsedError(string method, CentralServiceTransportResult transport, CentralServiceError error)
        {
            return new CentralServiceException(
                new CentralServiceError(
                    error.HttpStatus,
                    method,
                    transport.Url,
                    error.Kind,
                    AppendTransportContext(error.Message, transport),
                    error.ErrorCode,
                    error.ErrorKey,
                    error.RawBody));
        }

        private static string AppendTransportContext(string message, CentralServiceTransportResult transport)
        {
            var segments = new List<string>
            {
                $"端点={transport.BaseUrl}",
                $"尝试={transport.Attempt}/{transport.MaxAttempts}",
            };

            if (transport.SkippedEndpoints.Count > 0)
            {
                segments.Add("已跳过=" + string.Join("、", transport.SkippedEndpoints));
            }

            return string.Format("{0} ({1})", message ?? string.Empty, string.Join("; ", segments));
        }

        private static bool IsSuccess(HttpStatusCode code)
        {
            var value = (int)code;
            return value >= 200 && value <= 299;
        }
    }
}

File: SDK/CentralService/java/service/src/main/java/erp/centralservice/service/CentralServiceServiceClient.java
package erp.centralservice.service;

import erp.centralservice.service.errors.CentralServiceException;
import erp.centralservice.service.internal.CentralServiceErrorParser;
import erp.centralservice.service.internal.CentralServiceHttpClient;
import erp.centralservice.service.internal.CentralServiceJson;
import erp.centralservice.service.models.ApiResponse;
import erp.centralservice.service.models.ServiceRegistrationRequest;
import erp.centralservice.service.models.ServiceRegistrationResponse;

import java.util.Map;

/**
 * 面向服务注册侧的 Central Service SDK 客户端。
 */
public final class CentralServiceServiceClient {
    private final CentralServiceHttpClient http;

    public CentralServiceServiceClient(String baseUrl) {
        this(new CentralServiceSdkOptions(baseUrl));
    }

    public CentralServiceServiceClient(CentralServiceSdkOptions options) {
        if (options == null) throw new IllegalArgumentException("options is required");
        this.http = new CentralServiceHttpClient(options);
    }

    public ServiceRegistrationResponse register(ServiceRegistrationRequest request) {
        if (request == null) throw new IllegalArgumentException("request is required");

        CentralServiceHttpClient.Response response = send(
            "POST",
            "/api/Service/register",
            CentralServiceJson.stringify(request.toJson()));

        ApiResponse<ServiceRegistrationResponse> api = parseApiResponse(response.body, new ApiResponseDataParser<ServiceRegistrationResponse>() {
            @Override
            public ServiceRegistrationResponse parse(Object data) {
                return ServiceRegistrationResponse.fromJson(data);
            }
        });
        if (api == null) {
            throw new CentralServiceException(
                CentralServiceHttpClient.appendTransportContext(
                    CentralServiceErrorParser.unknown("POST", response.url, response.statusCode, response.body, "无法解析响应"),
                    response));
        }
        if (!api.success) {
            throw new CentralServiceException(
                CentralServiceHttpClient.appendTransportContext(
                    CentralServiceErrorParser.parse("POST", response.url, response.statusCode, response.body),
                    response));
        }
        return api.data;
    }

    public void deregister(String serviceId) {
        if (serviceId == null || serviceId.trim().isEmpty()) throw new IllegalArgumentException("serviceId is required");

        CentralServiceHttpClient.Response response = send(
            "DELETE",
            "/api/Service/deregister/" + CentralServiceHttpClient.encodePathSegment(serviceId),
            null);

        ApiResponse<Object> api = parseApiResponse(response.body, new ApiResponseDataParser<Object>() {
            @Override
            public Object parse(Object data) {
                return data;
            }
        });
        if (api != null && !api.success) {
            throw new CentralServiceException(
                CentralServiceHttpClient.appendTransportContext(
                    CentralServiceErrorParser.parse("DELETE", response.url, response.statusCode, response.body),
                    response));
        }
    }

    private CentralServiceHttpClient.Response send(String method, String path, String body) {
        final CentralServiceHttpClient.Response response;
        try {
            response = http.send(method, path, body);
        } catch (RuntimeException ex) {
            throw new CentralServiceException(CentralServiceHttpClient.createTransportError(method, path, ex));
        }

        if (!isSuccess(response.statusCode)) {
            throw new CentralServiceException(
                CentralServiceHttpClient.appendTransportContext(
                    CentralServiceErrorParser.parse(method, response.url, response.statusCode, response.body),
                    response));
        }
        return response;
    }

    private interface ApiResponseDataParser<T> {
        T parse(Object data);
    }

    private static <T> ApiResponse<T> parseApiResponse(String json, ApiResponseDataParser<T> dataParser) {
        if (json == null || json.trim().isEmpty()) return null;
        Object root = CentralServiceJson.parse(json);
        Map<String, Object> obj = CentralServiceJson.asObject(root);
        if (obj == null) return null;

        ApiResponse<T> api = new ApiResponse<T>();
        api.success = CentralServiceJson.asBoolean(obj.get("success"), false);
        api.errorCode = CentralServiceJson.asIntNullable(obj.get("errorCode"));
        api.errorMessage = CentralServiceJson.asString(obj.get("errorMessage"));
        api.data = dataParser == null ? null : dataParser.parse(obj.get("data"));
        return api;
    }

    private static boolean isSuccess(int statusCode) {
        return statusCode >= 200 && statusCode <= 299;
    }
}
